Remove array allocations routing in common cases

davidfowl opened the pull request:
While looking at some profiles these array allocations were showing up pretty high on the list once low hanging fruit was removed:
- The array of route values after matching
- The array of candidates matched for a route

Using C# 12's inline array feature, we can use stack allocated arrays in 2 cases:

1. When we have < 5 route values.
2. When we have < 5 route candidates (this is usually 1).

TODO: Check for throughput regressions. We have some BDN tests that I need to run.

Allocations:

Before

|Type|Allocations|Bytes|Average Size \(Bytes\)|
|-|-|-|-|
Microsoft.AspNetCore.Routing.Matching.CandidateState\[1\]|10,000|480,000|48|
System.Collections.Generic.KeyValuePair\<System.String, System.Object\>\[\]|10,000|400,000|40|
Microsoft.AspNetCore.Routing.RouteValueDictionary|10,000|400,000|40|

After

|Type|Allocations|Bytes|Average Size \(Bytes\)|
|-|-|-|-|
Microsoft.AspNetCore.Routing.RouteValueDictionary|10,000|1,119,400|111.94|

We've removed 2 heap allocations but increased the size of RVD (there are 4 extra KVP fields now). This was a request with a single route paramter.

davidfowl committed:
Make it work

davidfowl committed:
Use inline arrays for <= 4 route values

davidfowl committed:
It can't be null

davidfowl committed:
Fix comments

davidfowl committed:
Blank lines

stephentoub commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -23,14 +23,57 @@ namespace Microsoft.AspNetCore.Routing;
+    /// <returns>A new <see cref="RouteValueDictionary"/>.</returns>
+    public static RouteValueDictionary FromValues(Span<KeyValuePair<string, object?>> items)
+    {
+        // We need to compress the array by removing non-contiguous items. We
+        // typically have a very small number of items to process. We don't need
+        // to preserve order.
+        var start = 0;
+        var end = items.Length - 1;
+
+        // We walk forwards from the beginning of the array and fill in 'null' slots.
+        // We walk backwards from the end of the array end move items in non-null' slots
+        // into whatever start is pointing to. O(n)
+        while (start <= end)
+        {
+            if (items[start].Key != null)
+            {
+                start++;
+            }
+            else if (items[end].Key != null)
+            {
+                // Swap this item into start and advance
+                items[start] = items[end];
+                items[end] = default;
+                start++;
+                end--;
+            }
+            else
+            {
+                // Both null, we need to hold on 'start' since we
+                // still need to fill it with something.
+                end--;
+            }
+        }
+
+        return new RouteValueDictionary
+        {
+            _arrayStorage = new(items[..start]),
+            _count = start,
+        };
```
Could this instead be:
```C#
int next = 0;
for (int i = 0; i < items.Length; i++)
{
    if (items[i].Key is not null)
    {
        if (next != i) items[next] = items[i];
        next++;
    }
}

return new RouteValueDictionar
{
    _arrayStorage = new(items[..next]),
    _count = next,
}
```
?

davidfowl replied:
I didn't write the original code, just copied it, but I'll see if this works.

mavasani commented on src/Http/Http.Abstractions/src/Routing/DynamicArray.cs:
```diff
@@ -0,0 +1,143 @@
+    [InlineArray(DefaultCapacity)]
+    struct InlineArray<TItem>
+    {
+#pragma warning disable CA1823 // Avoid unused private fields
+#pragma warning disable IDE0044 // Add readonly modifier
+#pragma warning disable IDE0051 // Remove unused private members
+        private TItem _item0;
+#pragma warning restore IDE0051 // Remove unused private members
+#pragma warning restore IDE0044 // Add readonly modifier
+#pragma warning restore CA1823 // Avoid unused private fields
```
Can we file issues in both the Roslyn and roslyn-analyzers repo for the same?

CyrusNajmabadi replied:
tagging @Cosifne who will take these.

Cosifne replied:
And assigned the issue to me. I am the IDE side buddy to help with this.

stephentoub replied:
I opened https://github.com/dotnet/roslyn/issues/69143 and https://github.com/dotnet/roslyn-analyzers/issues/6789

davidfowl committed:
Fixed more tests

DamianEdwards commented on src/Http/Http.Abstractions/src/Routing/RouteValueDictionary.cs:
```diff
@@ -23,14 +23,57 @@ namespace Microsoft.AspNetCore.Routing;
+        var end = items.Length - 1;
+
+        // We walk forwards from the beginning of the array and fill in 'null' slots.
+        // We walk backwards from the end of the array end move items in non-null' slots
```
```suggestion
        // We walk backwards from the end of the array and move items in non-null' slots
```

davidfowl committed:
Remove more allocations
- Also remove array allocations when selecting the route
- Null out array in Clear in DynamicArray

davidfowl commented:
Made a part of the change in https://github.com/dotnet/aspnetcore/pull/49579 (the part that doesn't require public APIs)

davidfowl commented:
To make this change to `RouteValueDictionary` worthwhile, we need to pool or reuse the instances across requests. To be so far this feels like we're shuffling the cost around (moving from 2 references to 1) but not significantly reducing the overall set of allocations.
